server/server.go

```diff
@@ -6,53 +6,299 @@ package server
+	}
+
+	s.polling.ctx, s.polling.stop = context.WithCancel(context.Background())
+	s.startPolling()
+
+	return &s, nil
+}
+
+// GetAcceptedIndex atomically reads the accepted index
+func (serv *shipperServer) GetAcceptedIndex() int64 {
+	return atomic.LoadInt64(&serv.acceptedIndex)
+}
+
+// GetPersistedIndex atomically reads the persisted index
+func (serv *shipperServer) GetPersistedIndex() int64 {
+	return atomic.LoadInt64(&serv.persistedIndex)
+}
+
+// PublishEvents is the server implementation of the gRPC PublishEvents call.
+func (serv *shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
+	resp := &messages.PublishReply{
+		Uuid: serv.uuid,
+	}
+
+	// the value in the request is optional
+	if req.Uuid != "" && req.Uuid != serv.uuid {
+		resp.AcceptedIndex = serv.GetAcceptedIndex()
+		resp.PersistedIndex = serv.GetPersistedIndex()
+		serv.logger.
+			With(
+				"expected", serv.uuid,
+				"actual", req.Uuid,
+			).
+			Debugf("shipper UUID does not match, all events rejected")
+
+		return resp, nil
+	}
+
+	for _, e := range req.Events {
+		err := serv.queue.Publish(e)
+		if err == nil {
+			resp.AcceptedIndex++
+			continue
+		}
+
+		log := serv.logger.
+			With(
+				"event_count", len(req.Events),
+				"accepted_count", resp.AcceptedCount,
+			)
+
+		if errors.Is(err, queue.ErrQueueIsFull) {
+			log.Warnf("queue is full, not all events accepted")
+		} else {
+			err = fmt.Errorf("failed to enqueue an event: %w", err)
+			serv.logger.Error(err)
+		}
+
+		break
+	}
+
+	// it is cheaper to check than always using atomic
+	if resp.AcceptedCount > 0 {
+		atomic.AddInt64(&serv.acceptedIndex, int64(resp.AcceptedCount))
+	}
+	resp.AcceptedIndex = serv.GetAcceptedIndex()
+	resp.PersistedIndex = serv.GetPersistedIndex()
+
+	serv.logger.
+		With(
+			"event_count", len(req.Events),
+			"accepted_count", resp.AcceptedCount,
+			"accepted_index", resp.AcceptedIndex,
+			"persisted_index", resp.PersistedIndex,
+		).
+		Debugf("finished publishing a batch")
+
+	return resp, nil
+}
+
+// PublishEvents is the server implementation of the gRPC PersistedIndex call.
+func (serv *shipperServer) PersistedIndex(req *messages.PersistedIndexRequest, producer pb.Producer_PersistedIndexServer) error {
+	// first we reply with current values and then subscribe to notifications and start streaming
+	err := producer.Send(&messages.PersistedIndexReply{
+		Uuid:           serv.uuid,
+		PersistedIndex: serv.GetPersistedIndex(),
+	})
+	if err != nil {
+		return err
+	}
+
+	serv.logger.Debug("subscribing client for persisted index change notification...")
+	ch, stop, err := serv.notifications.subscribe()
+	if err != nil {
+		return err
+	}
+
+	serv.logger.Debug("client subscribed for persisted index change notifications")
+	// defer works in a LIFO order
+	defer serv.logger.Debug("client unsubscribed from persisted index change notifications")
+	defer stop()
+
+	for {
+		select {
+
+		case change := <-ch:
```

Comment:
```suggestion
		case change, closed := <-ch:
```

Should you be explicitly checking if the channel was closed so you don't send a zero-valued change back to a client when this happens?

Reply:
that's a point, however, I would have expected one of the contexts below to fire first.

I'll add the flag check anyway, better to be safe.

Reply:
> so you don't send a zero-valued change back to a client when this happens
btw this would not happen because of the next line :)